### ISSUE TITLE
add clang-tidy to Travis build

### ISSUE THREAD

**Issue by shalunov** [clostra/newnode#37]
- [x] try clang-tidy
- [x] check if it even works on C
- [x] post the output in this issue
- [x] figure out the relationship with #35
- [x] decide if it's worth it to add it to the build
- [x] add to Travis build

**Comment by inetic** [comment]
> try clang-tidy

Done

> check if it even works on C

It does.

> post the output in this issue

In the code we have mostly *.c files, but one `dht.cpp` file which acts as a C binding to the C++ project `libbtdht`. When running `clang-tidy` with the `dht.cpp` file there are plenty of warnings related to not using modern C++ techniques. I presume those are not important given the nature of the file (being a glue between C and C++) and after disabling plenty of them:

`-google-readability-casting,-llvm-include-order,-modernize-use-nullptr,-modernize-use-override,-misc-unused-parameter,-cppcoreguidelines-pro-type-cstyle-cast,-readability-implicit-bool-cast,-misc-unused-parameters,-modernize-use-auto,-cppcoreguidelines-pro-bounds-array-to-pointer-decay`

I'm left with these warnings:

```
/home/peter/work/dcdn/dht.cpp:65:5: warning: consider replacing 'long' with 'int64' [google-runtime-int]
    long fsize = ftell(f);
    ^
/home/peter/work/dcdn/dht.cpp:68:5: warning: do not use pointer arithmetic [cppcoreguidelines-pro-bounds-pointer-arithmetic]
    buf[fsize] = '\0'; // The BencEntity::Parse function expects this to be a null terminated string.
    ^
/home/peter/work/dcdn/dht.cpp:72:65: warning: do not use pointer arithmetic [cppcoreguidelines-pro-bounds-pointer-arithmetic]
    BencEntity::Parse((uint8_t *)buf, *ent, (unsigned char*)buf + fsize);
                                                                ^
/home/peter/work/dcdn/dht.cpp:119:9: warning: do not call c-style vararg functions [cppcoreguidelines-pro-type-vararg]
        printf("getaddrinfo: %s\n", gai_strerror(error));
        ^
/home/peter/work/dcdn/dht.cpp:215:28: warning: do not use pointer arithmetic [cppcoreguidelines-pro-bounds-pointer-arithmetic]
        buffer.assign(v, v + strlen(v));
                           ^
```

[Here](https://pastebin.com/jKbNw1SL) are the warning with all of the checks enabled.

This was generated with `clang-tidy-3.8`, interestingly, `clang-tidy-3.5` crashed. I only mention this because it may cause problems running `clang-tidy` on CI workers as those usually have only older versions of programs available.

With *.c files I again had to disable quite a few checks, but after doing that I did find one nasty buffer overflow:

```
/home/peter/work/dcdn/injector.c:88:9: warning: Excessive padding in 'proxy_request' (71 padding bytes, where 7 is optimal). Consider reordering the fields or adding explicit padding members [clang-analyzer-optin.performance.Padding]
typedef struct {
        ^
/home/peter/work/dcdn/injector.c:88:9: note: Excessive padding in 'proxy_request' (71 padding bytes, where 7 is optimal). Consider reordering the fields or adding explicit padding members
typedef struct {
        ^
Suppressed 121 warnings (121 in non-user code).
Use -header-filter=.* to display errors from all non-system headers.
62 warnings generated.
/home/peter/work/dcdn/injector_helper.c:560:20: warning: Access out-of-bound array element (buffer overflow) [clang-analyzer-alpha.security.ArrayBound]
    host[addr_len] = 0;
                   ^
/home/peter/work/dcdn/injector_helper.c:610:5: note: Taking true branch
    if (evhttp_request_get_command(req_in) == EVHTTP_REQ_CONNECT) {
    ^
/home/peter/work/dcdn/injector_helper.c:611:16: note: Calling 'create_tunnel'
        return create_tunnel(p, req_in);
               ^
/home/peter/work/dcdn/injector_helper.c:572:19: note: Calling 'parse_host_uri'
    bool parsed = parse_host_uri(uri, addr, sizeof(addr), &port);
                  ^
/home/peter/work/dcdn/injector_helper.c:543:5: note: Loop condition is true.  Entering loop body
    for (const char *c = uri; c != uri_end; ++c) {
    ^
/home/peter/work/dcdn/injector_helper.c:544:9: note: Taking false branch
        if (*c == ':') {
        ^
/home/peter/work/dcdn/injector_helper.c:543:31: note: Assuming 'c' is not equal to 'uri_end'
    for (const char *c = uri; c != uri_end; ++c) {
                              ^
/home/peter/work/dcdn/injector_helper.c:543:5: note: Loop condition is true.  Entering loop body
    for (const char *c = uri; c != uri_end; ++c) {
    ^
/home/peter/work/dcdn/injector_helper.c:544:9: note: Taking true branch
        if (*c == ':') {
        ^
/home/peter/work/dcdn/injector_helper.c:546:13: note:  Execution continues on line 550
            break;
            ^
/home/peter/work/dcdn/injector_helper.c:550:5: note: Taking false branch
    if (addr_end == uri_end) {
    ^
/home/peter/work/dcdn/injector_helper.c:560:20: note: Access out-of-bound array element (buffer overflow)
    host[addr_len] = 0;
                   ^
Suppressed 61 warnings (61 in non-user code).
Use -header-filter=.* to display errors from all non-system headers.
17 warnings generated.
Suppressed 17 warnings (17 in non-user code).
Use -header-filter=.* to display errors from all non-system headers.
63 warnings generated.
Suppressed 63 warnings (63 in non-user code).
Use -header-filter=.* to display errors from all non-system headers.
14 warnings generated.
/home/peter/work/dcdn/sha1.c:168:5: warning: Value stored to 'a' is never read [clang-analyzer-deadcode.DeadStores]
    a = b = c = d = e = 0;
    ^
/home/peter/work/dcdn/sha1.c:168:5: note: Value stored to 'a' is never read
    a = b = c = d = e = 0;
    ^
```

The disabled checks are:

[google-readability-casting](https://pastebin.com/pVyKjr0P)
[llvm-include-order](https://pastebin.com/Tz1NBfyY)
[clang-analyzer-alpha.core.CastToStruct](https://pastebin.com/DGqZXgtY)
[misc-unused-parameters](https://pastebin.com/BAkxMpzS)
[google-readability-todo](https://pastebin.com/rXgUmF8N)
[google-readability-braces-around-statements](https://pastebin.com/T8Js4aDp)
[readability-braces-around-statements](https://pastebin.com/13WzU3j9)
[misc-macro-parentheses](https://pastebin.com/dpC0Ae0W)

>  figure out the relationship with #35

[Clang-tidy is also a frontend for static analyzer](http://lists.llvm.org/pipermail/cfe-dev/2015-September/044966.html). From that link, static analyzer checks are those with the prefix `clang-analyzer-` and since I used the flag `-checks=*,-<excluded ones>`, they were used in my tests above (unless I explicitly excluded them).

**Comment by inetic** [comment]
> decide if it's worth it to add it to the build

I personally think it's useful, especially the `clang-analyzer-alpha.security.ArrayBound` analyzer which found the [bug](https://github.com/inetic/dcdn/commit/709c87a8e2ba25bfc6c7e03071c04f1e83561050). @shalunov, if you're of the same opinion, let me know which of the tests that produce warnings you prefer to stay excluded and for which of them I should modify the sources to remove warnings.

**Comment by shalunov** [comment]
The C++ warnings (don't use C) have a point in the general context, but not for the glue code, I'd say. 

Padding is the sort of thing that might be good to flag, but probably not break the build for. 

Unused initialized variables are a good find. Unfortunately, the really useful find, buffer overflow, is alpha and might be dependent on Clang version. Let's give it a try and see if the Travis version works. 

Can you try to minimize the diff, particularly of the build.sh file?

**Comment by shalunov** [comment]
I think you're meant to use `.clang-tidy` for clang-tidy options:

```
  -checks=<string>             -
                                 Comma-separated list of globs with optional '-'
                                 prefix. Globs are processed in order of
                                 appearance in the list. Globs without '-'
                                 prefix add checks with matching names to the
                                 set, globs with the '-' prefix remove checks
                                 with matching names from the set of enabled
                                 checks. This option's value is appended to the
                                 value of the 'Checks' option in .clang-tidy
                                 file, if any.
```

Looks like this way, only the compiler name would change in `build.sh`?

**Comment by inetic** [comment]
> I think you're meant to use .clang-tidy for clang-tidy options:

The problem with the `.clang-tidy` is that I'm having hard time finding documentation to all the options and its format. 

> Looks like this way, only the compiler name would change in build.sh?

There is still some difference, while compilation has the form:

```
clang <compile flags> -o executable <source c files> <include dirs> <static libs>
```
Clang-tidy has this format:

```
clang-tidy -checks='<checks>' <source c files> -- <compile flags> <include dirs>
```

[Here](https://github.com/clostra/dcdn/pull/41/files) is how I have it currently. Note that adding `.clang-tidy` config file would only help me remove [this  one line](https://github.com/clostra/dcdn/pull/41/files#diff-0b83f9dedf40d7356e5ca147a077acb4R75).

Now for the even worse news, `clang-tidy` isn't installed on the CI workers by default. I was mainly trying to add in on Linux one, but for that I had to add a new `clang` package source. That "kinda" worked and I got to even upgrade `clang` to versions 3.8 and 3.9 (default there is 3.4), but then the sanitization stopped working. I think the problem is that the `clang` from those external packages isn't as well polished and doesn't set paths to `Asan` libraries correctly. But I stopped investigating there.

The MacOS CI worker doesn't have `clang-tidy` neither. I tried `brew install llvm --with-clang --with-clang-extra-tools` as suggested [here](https://stackoverflow.com/questions/36496394/how-to-set-get-clang-build-on-travis-for-os-x) but with no joy.

I think this is becoming a can of worms. At the moment, the `build.sh` script [contains](https://github.com/clostra/dcdn/pull/41/files#diff-0b83f9dedf40d7356e5ca147a077acb4R73) an "if has `clang-tidy` then use it" clause so it's used locally but not on CI machines.

I've read somewhere that one could use a more recent Ubuntu on Travis, but that requires `docker`. This could solve some of our problems. The bad thing is that it doesn't work with MacOS (but perhaps it doesn't have to). I have zero experience with `docker` but I think @ivilata has some.
